More file extensions for Groff

larsbrinkhoff opened the pull request:
This PR adds more file extensions for the Groff language.  [Wikipedia](http://en.wikipedia.org/wiki/Man_page#Manual_sections) documents some of these, but I've also [estimated](https://github.com/search?q=extension%3A0+sh+pp) actual usage on GitHub:

| Extension | Files | Comment
| --- | --- | ---
| `.0` | 5112
| `.1in` | 1247
| `.1m` | 14193 | Sysadm tools
| `.1x` | 913 | X11 tools
| `.3in` | 11645
| `.3m` | 345 | Math libraries
| `.3qt` | 3518 | QT libraries
| `.3x` | 36119 | X11 libraries
| `.8` | 201,666 | Sysadm tools
| `.9` | 22718 | Syscalls
| `.me` | [3623](https://github.com/search?q=extension%3Ame++th+OR+pp+OR+nr+OR+"so+usr"&type=Code) | Using *me* macros
| `.ms` | 9930 | Using *ms* macros
| `.n` | 45067 | Tcl/Tk libraries
| `.rno` | [936](https://github.com/search?q=extension%3Arno+NOT+slartibartfast) |
| `.roff` | 909 |

The `.n` extension conflicts with `Nemerle`.  A new heuristic was needed to resolve this.  I verified with 1000 samples:

| | Before | After | Actual | Comment
| --- | --- | --- | --- | ---
| Groff | 8 | 950 | 910
| Binary | 25 | 25 | 25 | [Neko](http://nekovm.org/) VM bytecode
| Nemerle | 967 | 25 | 23
| Text | | | 24
| Font | | | 5
| Microcode | | | 4
| unknown | | | 4
| batchflow | | | 3
| HTML | | | 1
| Diff | | | 1

larsbrinkhoff commented:
@arfon @bkeepers @pchaigno Comments?

pchaigno commented:
Looks like Groff files only represent a small subset of all files for most of these extensions: 

Extension | Groff files | All files
----|----|----
`.0` | 5,112 | 594,549
`.1in` | 1,074 | 1,247
`.1m` | 9,801 | 14,442
`.1x` | 420 | 841
`.3in` | 11,091 | 11,645
`.3m` | 345 | 475
`.3qt` | 1,759 | 3,521
`.3x` | 36,119 | 36,357
`.8` | 146,607 | 246,089
`.9` | 19,877 | 59,741
`.me` | 404 | 15,702
`.ms` | 3,555 | 90,828
`.n` | 42,901 | 56,628
`.roff` | 665 | 1,115

larsbrinkhoff commented:
True for `.0`, `.ms`, and `.me`.  All the other are at least 50% of the total.

Also, in a previous PR, >600 files was enough to add `.f` as a Forth extension.  So I'm guessing a similar figure would apply here.

larsbrinkhoff commented:
A good chunk of the `.0` files are Unix `.so` files:
https://github.com/search?q=extension%3A0+filename%3Aso

pchaigno commented:
I update the table above to include Unix `.so` files and other languages I found for some of these extensions.

larsbrinkhoff commented:
@pchaigno Those "Text files related to Qt" are man pages.  `.so` is an include file directive.

larsbrinkhoff commented:
Looked into `.ms`.  In a set of 1000 samples, I find:

| Language | Files | Comment
| --- | --- | --- | --- | ---
| Assembly | 853 |
| Groff | 92 | 
| MaxScript | 19 | Scripting language for Autodesk 3ds Max
| Mustache | 15 | Ruby template library
| other | 3 | 

So I'll definately add `Assembly`.  Not sure if MaxScript or Mustache qualifies.

larsbrinkhoff commented:
Results after adding `.ms` to GAS:

| Language | Before | After | Actual
| --- | --- | --- | ---
| Assembly | | 853 | 853
| Groff | | 125 | 92
| MaxScript | | | 19
| Mustache | | | 15
| other | 1000 | 22 | 3

larsbrinkhoff commented:
Ping @arfon.  I believe I addressed @pchaigno's concerns:

- Removed `.0` because it may do more harm than good.
- "Text files related to Qt" are indeed `Groff`.
- Classify `read.me` as `Text`.
- Added `.ms` to `GAS`.

pchaigno commented:
> - Removed .0 because it may do more harm than good.
- "Text files related to Qt" are indeed Groff.
- Added .ms to GAS.

:+1: 
<br>

> Classify read.me as Text.

There are still a lot of other text files such as `click.me`, `delete.me`, etc.
We might need to wait for a way to classify file as "not a language" for this one... What do you think?

larsbrinkhoff commented:
> There are still a lot of other text files such as click.me, delete.me, etc.

Also, `readme.me`, `keep.me`, `test.me`.  But some of those files are empty (zero length).  What language are those?

But, how many files are there?  I got the excact same number of files for many searches.  That can't be right.

> We might need to wait for a way to classify file as "not a language" for this one... What do you think?

Yes, that has come up many times before.

pchaigno commented:
> What language are those?

Most of them seem to be text files: [`delete.me`](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ame+filename%3Adelete&type=Code&ref=searchresults), [`read.me`](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ame+filename%3Aread&type=Code&ref=searchresults), [`click`](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ame+filename%3Aclick&type=Code&ref=searchresults), etc.
<br>

> Yes, that has come up many times before.

It should arrive with the Google Summer of Code I think (@lazywei ?).

larsbrinkhoff commented:
E.g. `delete.me` are mostly empty files:

https://github.com/search?q=extension%3Ame+filename%3Adelete+NOT+default+NOT+placeholder+NOT+delete

Not sure they can be called text files?

lazywei commented:
I hope we can deliver that feature in our GSoC work as well. However, for now, we're focusing on re-write the classifier from Ruby to Go (WIP https://github.com/lazywei/mockingbird). Hopefully after the porting, we can add more features such like classify file as "not a language".

larsbrinkhoff commented:
@arfon I pulled out the `.me` changes (separate PR forthcoming).  Does the rest look ok?

pchaigno commented:
Looks :+1: to me. @arfon ?

larsbrinkhoff commented:
Found one more extension: `.rno`.

larsbrinkhoff commented:
Ping @arfon  Please look at this?

arfon commented:
:+1: this looks solid. Thanks for the monumental effort @larsbrinkhoff!

Alhadis commented:
> Not sure if MAXScript qualifies.

Well, it's not passing JavaScript in popularity... but it *is* a legitimate language with enough users to warrant recognition.

I was surprised to see my [local setup](https://github.com/Alhadis/3DSMax-Setup) flagged as being 87% *ROFF from all the `.ms` extensions. I'll probably look into submitting a pull request for the language later... unless somebody here deems it too unimportant to qualify...? @arfon , would it be okay?
